File: backend/app/schemas/forum.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator


ForumCategoryType = Literal["movie", "series", "book", "custom"]


class ForumThreadCreate(BaseModel):
    title: str = Field(min_length=3, max_length=200)
    text: str = Field(min_length=1, max_length=5000)
    category_type: ForumCategoryType
    custom_category: str | None = Field(default=None, min_length=2, max_length=50)

    @model_validator(mode="after")
    def validate_category(self):
        if self.category_type == "custom":
            if not self.custom_category or not self.custom_category.strip():
                raise ValueError("custom_category is required when category_type='custom'")
            self.custom_category = self.custom_category.strip()
        else:
            self.custom_category = None
        return self


class ForumThreadUpdate(BaseModel):
    title: str = Field(min_length=3, max_length=200)
    text: str = Field(min_length=1, max_length=5000)
    category_type: ForumCategoryType
    custom_category: str | None = Field(default=None, min_length=2, max_length=50)

    @model_validator(mode="after")
    def validate_category(self):
        if self.category_type == "custom":
            if not self.custom_category or not self.custom_category.strip():
                raise ValueError("custom_category is required when category_type='custom'")
            self.custom_category = self.custom_category.strip()
        else:
            self.custom_category = None
        return self


class ForumPostCreate(BaseModel):
    text: str = Field(min_length=1, max_length=5000)

    # Старе поле залишаємо для сумісності з поточним фронтом.
    # Backend буде сприймати його як "пост, на який натиснули Reply".
    parent_post_id: str | None = None

    # Нове бажане поле для нової логіки.
    # Це конкретний пост, на який відповідаємо.
    reply_to_post_id: str | None = None

class ForumPostUpdate(BaseModel):
    text: str = Field(min_length=1, max_length=5000)


class ForumThreadResponse(BaseModel):
    id: str
    user_id: str
    author_username: str
    author_avatar_id: str
    title: str
    text: str
    category_type: ForumCategoryType
    custom_category: str | None
    score: int
    replies_count: int
    created_at: datetime
    updated_at: datetime
    last_activity_at: datetime
    edited: bool


class ForumPostBaseResponse(BaseModel):
    id: str
    thread_id: str
    user_id: str
    author_username: str
    author_avatar_id: str
    text: str
    score: int

    # ID головного поста, під яким лежить відповідь.
    # Для top-level post буде None.
    parent_post_id: str | None

    # ID конкретного поста, на який натиснули Reply.
    reply_to_post_id: str | None = None

    # ID користувача, якому відповіли.
    reply_to_user_id: str | None = None

    # Username користувача, якому відповіли.
    reply_to_username: str | None = None

    created_at: datetime
    updated_at: datetime
    edited: bool


class ForumPostReplyResponse(ForumPostBaseResponse):
    pass


class ForumPostResponse(ForumPostBaseResponse):
    replies: list[ForumPostReplyResponse] = Field(default_factory=list)


class ForumActionResponse(BaseModel):
    message: str

File: frontend/src/types/forum.ts
export type ForumCategoryType = 'movie' | 'series' | 'book' | 'custom'
export type ForumThreadSort = 'activity' | 'newest' | 'score'

export type ForumThreadResponse = {
  id: string
  user_id: string
  author_username: string
  author_avatar_id: string
  title: string
  text: string
  category_type: ForumCategoryType
  custom_category: string | null
  score: number
  replies_count: number
  created_at: string
  updated_at: string
  last_activity_at: string
  edited: boolean
}

export type ForumThreadCreatePayload = {
  title: string
  text: string
  category_type: ForumCategoryType
  custom_category: string | null
}

export type ForumPostBaseResponse = {
  id: string
  thread_id: string
  user_id: string
  author_username: string
  author_avatar_id: string
  text: string
  score: number
  parent_post_id: string | null
  reply_to_post_id: string | null
  reply_to_user_id: string | null
  reply_to_username: string | null
  created_at: string
  updated_at: string
  edited: boolean
}

export type ForumPostReplyResponse = ForumPostBaseResponse

export type ForumPostResponse = ForumPostBaseResponse & {
  replies: ForumPostReplyResponse[]
}

export type ForumPostCreatePayload = {
  text: string
  reply_to_post_id?: string
}

export type ForumVotePayload = {
  value: 1 | -1
}

export type ForumVoteResponse = {
  target_type: 'thread' | 'post'
  target_id: string
  current_vote: 1 | -1 | null
  score: number
  message: string
}

export type ForumMyVoteResponse = {
  target_type: 'thread' | 'post'
  target_id: string
  current_vote: 1 | -1 | null
}